### ISSUE TITLE
Feature Request: Allow users to upload tides for tide correction

### ISSUE THREAD

**Issue by 2320sharon** [SatelliteShorelines/CoastSeg#281]
Some regions of the world have different tide levels that the included FES2014 model may not perform well at. To remedy this CoastSeg should allow the user to upload a CSV file containing the tides. 

| Transect ID   | Tide | Datetime |
| -------- | ------- | ---------|
| 1          |  2        | 2021-04-05 14:49:58+00:00   |
| 2         | 1       |2021-04-07 14:49:58+00:00   |4
| 2         | 1.2      |2021-04-07 18:49:58+00:00   |
| 3    | 1.5   |  2021-04-06 14:49:58+00:00|

**Comment by 2320sharon** [comment]
Here are the possible forms that uploaded tides can be accepted in

I heard back from our team and they mentioned they wanted the format dates x transect_ids as a format as well.

Below is an updated list of all the possible formats we should expect

1. columns : tide, date
    - Use the tide for each matching transect id

| Tide  | Date       |
|-------|------------|
| 1.2   | 2021-04-05 |
| 1.4   | 2021-04-06 |
| 1.3   | 2021-04-07 |


2. columns : transect id, date tide
  -  use tide for everything up to and including the date for that transect
 - If only one transect id is listed then that slope is used for tide correction for all the slopes
 
| Transect ID | Date       | Tide  |
|-------------|------------|-------|
| 1           | 2021-04-05 | 1.2   |
| 2           | 2021-04-05 | 1.4   |
| 2           | 2021-04-08 | 1.5   |
| 3           | 2021-04-06 | 1.3   |



3. columns : slope, x,y
   - Use the slope for the transect closet to the x and y location

| Date       | Tide | x       | y        |
|------------|------|---------|----------|
| 2021-04-05 | 1.2  | 34.0522 | -118.2437|
| 2021-04-06 | 1.4  | 36.7783 | -119.4179|
| 2021-04-07 | 1.3  | 37.7749 | -122.4194|


4. dates x transect ids.
- Basically its a pivot of form 2

| 123 | 1234 |12345 | 123456 |
| -------- | ------- | ---------| ---------|
| 2004-04-07 00:00:00+00:00          |  0.05                   | 0.04     | 0.04     |
|2004-04-08 00:00:00+00:00       | 0.04       |0.05      | 0.08     |
| 2004-04-09 00:00:00+00:00      | 0.03     |0.09    |0.07   |
| 2004-04-10 00:00:00+00:00     | 0.02   |  0.01  | 0.07      |

**Comment by 2320sharon** [comment]
This is related to the formats in issue https://github.com/SatelliteShorelines/CoastSeg/issues/282

This issue is related to https://github.com/SatelliteShorelines/CoastSeg/issues/264

**Comment by 2320sharon** [comment]
- [ ] Allow users to to specify a default slope value if one of the transects provided does not have any slope values
- [ ] Make functions to read the tides in all 4 formats
- [ ] Make tide correction work with tides for all 4 formats
- [ ] Ensure tide and slope values are saved out to the tidally_corrected_timeseries_merged csv file

**Comment by 2320sharon** [comment]
- [ ] Raise exception telling user that none of the transect ids matched if none of the uploaded transect ids matched

**Comment by 2320sharon** [comment]
- [ ] Make functions to read the tides in all 4 formats
The code below can read in the slopes/tides in all 4 formats

```


def compute_distance_xy(x1, y1, x2, y2):
    return np.sqrt((x1-x2)**2 + (y1-y2)**2)


def _initialize_column(df, column_name):
    """Initialize a new column with NaN values if it doesn't exist."""
    if column_name not in df.columns:
        df[column_name] = np.nan
    return df

def _find_closest_date(date, reference_dates):
    """Find the closest date from a series of dates."""
    return reference_dates.iloc[(reference_dates - date).abs().argsort()[0]]

# Only cares about column name and 'dates' column
def match_via_date(timeseries, df, column_name):
    """
    Match values based on closest date.
    
    Args:
        timeseries: DataFrame containing shoreline data
        df: Reference DataFrame with matching data
        column_name: Name of column to match
    
    Returns:
        Updated timeseries DataFrame with matched values
    """
    timeseries = _initialize_column(timeseries, column_name)
    
    # Loop through shoreline dates
    for date in timeseries['dates']:
        closest_date = _find_closest_date(date, df['dates'])
        timeseries.loc[timeseries['dates'] == date, column_name] = (
            df.loc[df['dates'] == closest_date, column_name].values[0]
        )
    
    return timeseries

# Only cares about column name, 'transect_id', and 'dates' column
def match_via_id_and_date(timeseries, df, column_name):
    """
    Match values based on transect_id and closest date.
    
    Args:
        timeseries: DataFrame containing shoreline data
        df: Reference DataFrame with matching data
        column_name: Name of column to match
    
    Returns:
        Updated timeseries DataFrame with matched values
    """
    timeseries = _initialize_column(timeseries, column_name)
    unique_ids = np.unique(df['transect_id'])
    
    for transect_id in unique_ids:
        matching_rows = df.loc[df['transect_id'] == transect_id]
        
        if len(matching_rows) == 1:
            # Single tide value case
            matching_value = matching_rows[column_name].values[0]
            timeseries.loc[timeseries['transect_id'] == transect_id, column_name] = matching_value
        elif len(matching_rows) > 1:
            # Multiple dates case
            shoreline_dates = timeseries.loc[timeseries['transect_id'] == transect_id, 'dates']
            for shoreline_date in shoreline_dates:
                closest_row = _find_closest_date(shoreline_date, matching_rows['dates'])
                matching_value = closest_row[column_name].values[0]
                timeseries.loc[
                    (timeseries['transect_id'] == transect_id) & 
                    (timeseries['dates'] == shoreline_date), 
                    column_name
                ] = matching_value
    
    return timeseries

# Only cares about column name, 'transect_id', and 'dates' column
def match_via_points_and_date(timeseries, df, column_name):
    """
    Matches tide measurements to transects based on closest spatial and temporal proximity.
    
    Parameters:
    ----------
    timeseries : DataFrame
        Contains transect data with columns: 'transect_id', 'x', 'y', 'dates'
    df : DataFrame
        Contains tide data with columns: 'x', 'y', 'dates', column_name
    column_name : str
        Name of the column in df containing the values to match
        
    Returns:
    -------
    DataFrame
        Input timeseries DataFrame with added column_name containing matched values
    """
    # Get unique transect IDs
    df_transect_id = np.unique(timeseries['transect_id'])
    
    for transect_id in df_transect_id:
        # Get transect coordinates
        transect_x, transect_y = timeseries.loc[timeseries['transect_id']==transect_id, ['x', 'y']].values[0]
        
        # Calculate distances to all tide points
        df['distance'] = compute_distance_xy(
            transect_x, 
            transect_y, 
            df['x'], 
            df['y']
        )
        
        # Find points with minimum distance
        min_distance = np.min(df['distance'])
        closest_tides = df.loc[df['distance']==min_distance]
        
        # For each date in this transect, find the closest tide temporally
        for date in timeseries.loc[timeseries['transect_id']==transect_id, 'dates']:
            closest_date = closest_tides.iloc[(closest_tides['dates']-date).abs().argsort()[:1]]
            value = closest_date[column_name].values[0]
            
            # Assign the value to the matching transect and date
            timeseries.loc[
                (timeseries['transect_id']==transect_id) & 
                (timeseries['dates']==date), 
                column_name
            ] = value
            
    return timeseries

def melt_df(df,column_name):
    df.rename(columns = {'Unnamed: 0':'dates'}, inplace = True)
    df['dates'] = pd.to_datetime(df['dates'])
    df = pd.melt(df, id_vars=['dates'], var_name='transect_id', value_name=column_name)
    return df

def read_content_csv(tides_file,timeseries,column_name='tide'):
    """
    Read tide data from a CSV file.
    
    Args:
        tides_file: Path to the tide data CSV file
        timeseries: DataFrame containing shoreline data
        column_name: Name of column to match
    
    Returns:
        DataFrame containing tide data
    """
    tides_df = pd.read_csv(tides_file)
    tides_df['dates'] = pd.to_datetime(tides_df['dates'])

    # if it has columns 'transect_id', 'tide', 'dates'
    if 'transect_id' in tides_df.columns:
        merged_csv = match_via_id_and_date(timeseries, tides_df, column_name)
    # if it has columns 'x', 'y', 'tide', 'dates'
    elif 'x' in tides_df.columns:
        merged_csv = match_via_points_and_date(timeseries, tides_df, column_name)
    # if it has columns 'dates', 'tide'
    elif 'dates' in tides_df.columns and column_name in tides_df.columns:
        merged_csv = match_via_date(timeseries, tides_df, column_name)
    # if one of column is called unnamed then we need to melt
    elif 'Unnamed: 0' in tides_df.columns:
        tides_df = melt_df(tides_df,column_name)
        merged_csv = match_via_id_and_date(timeseries, tides_df, column_name)
    else:
        raise ValueError('CSV format not supported. ')
    return merged_csv


```

**Comment by 2320sharon** [comment]
This code is my work in progress for both #282  and this issue. These new tide  correction functions that have `_new` appended to the function name would replace the existing functions. These new tide correction functions would take the `tides_file` as an optional parameter and the `beach slopes` can be passed as either a float or file.
[new_tide_corrections.zip](https://github.com/user-attachments/files/18677744/new_tide_corrections.zip)

**Comment by 2320sharon** [comment]
I probably want to overall the predict tides script to make tide predictions in one of these 4 formats as well. That would make it easier for users to do tide correction and take these tide predictions elsewhere.

**Comment by 2320sharon** [comment]
here is a work in progress for the new controls to upload tides or use the tide model as well as the option to use a static slope or upload a slopes file

![Image](https://github.com/user-attachments/assets/828713df-3315-4978-a818-d602d29fe012)

**Comment by 2320sharon** [comment]
Added guide for tide file formats and linked to it the widget: https://satelliteshorelines.github.io/CoastSeg/tide-file-format/

**Comment by 2320sharon** [comment]
Make a 5th format
seasonal slopes

match a transect id and month to a slope.

- [ ] match a transect id and month to a slope.

**Comment by 2320sharon** [comment]
Over Slack today Dan mentioned that instead of not tide correcting transects that didn't have a slope we should use the median transect instead. Add this to the process of reading the slopes from CSV and add a line to the user interface that says the median slope will be used for transects that were missing slope values.

- [ ] update UI_elements.py with a warning that says "median slope will be used for transects missing slope values"
- [ ] compute the median slope when reading slopes from CSV file if any of the transect IDs are missing from the CSV file

**Comment by 2320sharon** [comment]
- [ ] Update elevation text to read "Reference elevation (m, relative to user-specified vertical datum)" in both notebooks
- [ ] Update elevation text to read "Reference elevation (m, relative to user-specified vertical datum)" in all scripts

**Comment by 2320sharon** [comment]
- [ ] Update the website for tide correction to note that this corrects the shoreline by `shoreline + ((tide-reference_elevation)/beach_slope)`

**Comment by 2320sharon** [comment]
This feature has been added in CoastSeg v1.4.3
